src/lib/libast/man/error.3
.TH ERROR 3
.SH NAME
error \- error and debug trace message formatter
.SH SYNOPSIS
.nf
.ft CR
#include <error.h>

Error_info_t error_info;

void         error(int \fIlevel\fP, ...);
void         errorv(const char* \fIlibrary\fP, int \fIlevel\fP, va_alist \fIargs\fP);
int          errormsg(const char* \fIdictionary\fP, int \fIlevel\fP, ...);

#include <debug.h>

debug(\fIstatement\fP)
message((int \fIlevel\fP, ...))
messagef((void* \fIhandle\fP, void* \fIdiscipline\fP, int \fIlevel\fP, ...))
.ft
.fi
.SH DESCRIPTION
\f[CB]error\fP
is the error and debug trace message formatter.
.I level
is the severity level.
Messages with
.I "level < error_info.trace"
are suppressed.
.I error_info.trace
is initially
\f[CB]0\fR.
The remaining arguments are passed on to
\f[CB]printf\fR.
A
.I newline
is appended to the message text, so none should appear in the
\f[CB]printf\fP
format.
If
.I error_info.id
is not
\f[CB]0\fP
then messages with
.I "level > 0"
are prefixed by
.IR error_info.id: .
.PP
Before the message text is output to standard error
it is passed to the function
\f[CB]char* ERROR_translate(const char* \fItext\fP, int \fIflag\fP)\fR.
By default
\f[CB]ERROR_translate\fP
returns the
\f[CB]text\fP
argument, but on some systems it may do language translation via lookup
on the original source text.
\fR(\fP error
calls
\f[CB]ERROR_translate\fP
with a 0
\f[CB]flag\fP
argument).
.PP
.I level
may be one of:
.TP
.B <0
Negative values are for debug tracing.
Debug messages are prefixed with
.BI debug level.
If
.I "errno != error_info.last_errno"
then
.I error_info.last_errno
is set to
.I errno
and the error text for errno is appended to the message.
.TP
.B "ERROR_INFO [0]"
Information only; no prefixes are added to the message.
.TP
.B "ERROR_WARNING [1]"
\f[CB]warning:\fP
is added after
\f[CB]error_info.id\fP
and
.I error_info.warnings
is incremented.
.TP
.B "ERROR_ERROR [2]"
(soft error)
.I error_info.errors
is incremented.
.TP
.B ">= ERROR_FATAL [3]"
(hard error)
.I error_info.errors
is incremented and
\f[CB]exit(\fIlevel\fP\-2)\fP
is called after the message is emitted.
.TP
.B "ERROR_PANIC [130]"
(unrecoverable internal error)
\f[CB]panic:\fP
is added after
.IR error_info.id .
.PP
The following may be inclusive-or'd into
.I level
for alternate behavior:
.TP
.B ERROR_SYSTEM
The error text for
.I errno
is appended to the message.
.TP
.B ERROR_OUTPUT
The next argument is the file descriptor where the error message
should be emitted.
.TP
.B ERROR_SFIO_OUT
The next argument is the Sfio stream where the error message
should be emitted.
.TP
.B ERROR_SOURCE
Then next two arguments are a file name and line number that are added
to the message after
.IR error_info.id .
.TP
.B ERROR_USAGE
A usage message is emitted.
.TP
.B ERROR_PROMPT
The trailing
.I newline
is suppressed.
.TP
.B ERROR_NOID
The
.I error_info.id
prefix is suppressed.
.TP
.B ERROR_LIBRARY
The message is from a library routine.
.SH ENVIRONMENT
The elements of the global struct
.I error_info
control error output and actions.
Parts of
.I error_info
can be initialized from the
\f[CB]ERROR_OPTIONS\fP
environment variable.
\f[CB]ERROR_OPTIONS\fP
contains space separated
.IR name [ =value ]
options, described below.
.TP
.I "int core"
If
.I "error_info.core != 0"
then
.I "level >= error_info.core"
generates a core dump.
Initialized by
.nf
.ft CR
ERROR_OPTIONS="core=\fIlevel\fP"
.ft
.fi
where
.I level
can be a number or one of
\f[CB]error\fR,
\f[CB]fatal\fR,
or
\f[CB]panic\fR.
.I error_info.core
is a handy way to get a stack trace at the exact point of error.
.TP
.I "int error_info.trace"
If
.I "error_info.trace != 0"
and
.I "level < error_info.trace"
then the error message text is suppressed.
\f[CB]exit()\fP
may still be called if appropriate for
.IR level .
Initialized by
.nf
.ft CR
ERROR_OPTIONS="trace=\fIlevel\fP"
.ft
.fi
where
.I error_info.trace
is set to the negative of
.IR level .
.PP
Library error messages, suppressed by default, are enabled by
.nf
.ft CR
ERROR_OPTIONS="library"
.ft
.fi
The system
.I errno
message text can be forced for each message by
.nf
.ft CR
ERROR_OPTIONS="system"
.ft
.fi
.SH "EXTENDED DESCRIPTION"
\f[CB]<debug.h>\fP
provides debugging message macros when
\f[CB]DEBUG\fP
or
\f[CB]_BLD_DEBUG\fP
are defined
\fR(\fP _BLD_DEBUG
is defined by
the
.IR Mamfile s
when the
\f[CB]\-G\fP
a.k.a.
\f[CB]\-\-debug\-symbols\fP
option is passed to
.BR mamake (1),
which the
.BR package (1)
command does automatically when given the argument
\f[CB]debug=1\fR).
All of the macros expand to nothing when both
\f[CB]DEBUG\fP
and
\f[CB]_BLD_DEBUG\fP
are not defined.
Otherwise
\f[CB]debug\fP
expands its arg and
\f[CB]messagef\fP
and
\f[CB]message\fP
call
\f[CB]errorf\fP
and
\f[CB]error\fP
respectively if
.IR "error_info.trace<0" .
Notice that
\f[CB]messagef\fP
and
\f[CB]message\fP
are macro hacks that require double parentheses ((...)) around the
arguments.
.SH EXAMPLE
To enable debugging message level -3, library messages, and system
.I errno
text for all commands:
.nf
.ft CR
export ERROR_OPTIONS="trace=3 library system"
.ft
.fi


NEWS
This documents significant changes in the dev branch of ksh 93u+m.
For full details, see the git log at: https://github.com/ksh93/ksh
Uppercase BUG_* IDs are shell bug IDs as used by the Modernish shell library.

2026-09-23:

- Fixed failures (assignment data loss) when a nested associative array was
  assigned to a name reference that pointed to a variable in a calling
  function's local scope.

- The sleep builtin will now reject values beyond UINT_MAX and negative numbers.

- Fixed bugs that caused many builtins (such as kill) to mishandle out of
  range numbers.

2026-09-21:

- Fixed a bug in nested indexed array assignments where values that happened
  to be the names of declaration commands would be executed as those
  declaration commands. For example, arr=((a b c)(typeset -p PATH)) now
  works correctly by assigning 'typeset', '-p' and 'PATH' as array values to
  arr[1][0], arr[1][1] and arr[1][2] instead of executing 'typeset -p PATH'.

- Fixed a buffer overflow that could occur when creating a type with
  'typeset -T', creating a function, then attempting to use 'typeset -p'.

2026-09-19:

- Fixed a longstanding bug where already-assigned associative array elements
  cannot be used in default assignment values of another element of same
  array -- for example, ${x[1]:=${x[0]}} failed to work correctly.

- Fixed default assignments on unset associative array elements, e.g.
  ${x[sub]=value} or ${x[sub]:=value}, which failed to assign the value to
  the element and instead left it empty or nonexistent.

- Fixed a bug in unsetting an associative array where the values of the array
  would be cleared but the former array variable still had leftover attributes.

2026-09-17:

- Fixed multiple memory leaks occurring upon subshell exit when assigning a
  value to a variable with a shell discipline function within a subshell.

- Fixed a crash occurring when assigning, within a subshell, an invalid
  value to a variable of a type defined by the 'enum' command.

2026-09-16:

- Fixed: when the special IFS (field splitting characters) variable was
  assigned to within a ${ shared-state command substitution }, the state of IFS
  failed to persist beyond it, and a use after free (potential crash) occurred.

- Fixed bounds checking when assigning to fixed-size arrays (declared like,
  e.g., 'typeset -a array[4]'). Assigning to an out-of-range subscript now
  produces the expected error instead of potentially crashing.

2026-09-15:

- Fixed a memory leak when assigning, within a subshell, a value to a
  special variable (one of IFS, FPATH, CDPATH, SHELL, ENV, VISUAL, EDITOR,
  HISTFILE, OPTIND, SECONDS, RANDOM,.sh.match, .sh.version, LINENO, _,
  LC_CTYPE, LC_NUMERIC, LC_COLLATE, LC_MESSAGES, LC_ALL, and LANG).

2026-09-13:

- Fixed a memory leak that occurred when converting an existing scalar
  variable to an associative array using 'typeset -A'.

2026-09-12:

- Implemented advisory locking for $HISTFILE (~/.sh_history by default) so
  that multiple ksh instances can write to the history file without risking
  data loss when one of them decides to trim the file. Note that ksh
  versions before 2026-09-12 will ignore the lock, so the risk continues to
  exist if an older ksh binary shares write access to the file.

- History file trimming no longer requires write access to the directory in
  which the history file resides and will no longer change the directory's
  last-modified timestamp. However, it now requires a usable /tmp or $TMPDIR.

- I/O errors while writing to the history file are no longer silently ignored.

- Fixed a possible crash when attempting to set an integer variable to a
  multibyte character.

2026-09-10:

- Fixed a rare crash that could occur if a signal (such as interrupt, Ctrl+C)
  occurred while cleaning up the state of a virtual subshell upon its exit.

2026-09-08:

- Fixed: PWD and ENV were not properly reinitialised when executing a shell
  script without a #! path.

- Fixed a possible crash after unsetting RANDOM. Bug introduced on 2022-07-21.

2026-09-04:

- Fixed a longstanding bug in which output data loss or corruption could
  intermittently occur if, within a command substitution, standard error is
  redirected to standard output and, within that redirection, standard
  output is then connected to a pipe.

2026-09-02:

- Fixed a bug, introduced on 2025-05-30, that made 'print -s' ineffective
  at writing to the history file. This functionality is now restored.

2026-08-31:

- Fixed a crash in the expr(1) path-bound built-in command that could occur
  when using the non-standard 'match' or 'index' keyword.

2026-08-30:

- Fixed a bug that caused non-importable environment variables (i.e., those
  with names that are not valid shell variable names) to disappear from the
  environment of child processes after invoking more than two external
  commands, potentially crashing the shell. Bug introduced on 2025-05-30.

2026-08-29:

- If an error (such as a syntax error) occurs in a trap action, the error
  message will now have the word "(trap)" in the call stack for clarity.

2026-08-26:

- Following up to 2026-08-16, fixed a remaining problem in the 'read'
  built-in with combined -d and -n options when reading from a pipe; the
  custom delimiter is now correctly used instead of newline.

2026-08-16:

- Fixed a long-standing bug in the 'read' built-in: the -d option was
  ignored when combined with -n. The two options can now be used together.

2026-08-14:

- In the --man/--help output for built-in commands, text marked as italic
  is no longer rendered as bold + underlined, but only as underlined.

2026-08-13:

- Fixed corruption of $_, crashing the shell upon 'typeset -p _', after an
  error occurs trying to redefine a type defined with 'typeset -T'.

2026-08-06:

- Fixed ksh crashing upon parsing a very long list of simple commands
  separated by semicolons or, in 'eval' or dot script contexts, by newlines.


src/lib/libast/misc/error.c
/***********************************************************************
*                                                                      *
*               This software is part of the ast package               *
*          Copyright (c) 1985-2011 AT&T Intellectual Property          *
*          Copyright (c) 2020-2026 Contributors to ksh 93u+m           *
*                      and is licensed under the                       *
*                 Eclipse Public License, Version 2.0                  *
*                                                                      *
*                A copy of the License is available at                 *
*      https://www.eclipse.org/org/documents/epl-2.0/EPL-2.0.html      *
*         (with md5 checksum 84283fa8859daf213bdda5a9f8d1be1d)         *
*                                                                      *
*                 Glenn Fowler <gsf@research.att.com>                  *
*                  David Korn <dgk@research.att.com>                   *
*                   Phong Vo <kpv@research.att.com>                    *
*                  Martijn Dekker <martijn@inlv.org>                   *
*            Johnothan King <johnothanking@protonmail.com>             *
*                                                                      *
***********************************************************************/
/*
 * Glenn Fowler
 * AT&T Research
 *
 * error and message formatter
 *
 *	level is the error level
 *	level >= error_info.core!=0 dumps core
 *	level >= ERROR_FATAL calls error_info.exit
 *	level < 0 is for debug tracing
 *
 * NOTE: id && ERROR_NOID && !ERROR_USAGE implies format=id for errmsg()
 */

#include "lclib.h"

#include <ctype.h>
#include <ccode.h>
#include <namval.h>
#include <sig.h>
#include <stk.h>
#include <times.h>
#include <regex.h>

/*
 * 2007-03-19 move error_info from _error_info_ to (*_error_infop_)
 *	      to allow future Error_info_t growth
 */

static Error_info_t	_error_info_ =
{
	2, exit, write,
	0,0,0,0,0,0,0,0,
	0,			/* version			*/
	0,			/* auxiliary			*/
	0,0,0,0,0,0,0,		/* top of old context stack	*/
	0,0,0,0,0,0,0,		/* old empty context		*/
	0,			/* time				*/
	translate,
	0			/* catalog			*/
};
Error_info_t*		_error_infop_ = &_error_info_;

/*
 * these should probably be in error_info
 */

static struct State_s
{
	char*		prefix;
	Sfio_t*		tty;
	unsigned long	count;
	int		breakpoint;
	regex_t*	match;
} error_state;

#undef	ERROR_CATALOG
#define ERROR_CATALOG	(ERROR_LIBRARY<<1)

#define OPT_BREAK	1
#define OPT_CATALOG	2
#define OPT_CORE	3
#define OPT_COUNT	4
#define OPT_FD		5
#define OPT_LIBRARY	6
#define OPT_MASK	7
#define OPT_MATCH	8
#define OPT_PREFIX	9
#define OPT_SYSTEM	10
#define OPT_TIME	11
#define OPT_TRACE	12

static const Namval_t		options[] =
{
	"break",	OPT_BREAK,
	"catalog",	OPT_CATALOG,
	"core",		OPT_CORE,
	"count",	OPT_COUNT,
	"debug",	OPT_TRACE,
	"fd",		OPT_FD,
	"library",	OPT_LIBRARY,
	"mask",		OPT_MASK,
	"match",	OPT_MATCH,
	"prefix",	OPT_PREFIX,
	"system",	OPT_SYSTEM,
	"time",		OPT_TIME,
	"trace",	OPT_TRACE,
	0,		0
};

/*
 * called by stropt() to set options
 */

static int
setopt(void* a, const void* p, int n, const char* v)
{
	NoP(a);
	if (p)
		switch (((Namval_t*)p)->value)
		{
		case OPT_BREAK:
		case OPT_CORE:
			if (n)
				switch (*v)
				{
				case 'e':
				case 'E':
					error_state.breakpoint = ERROR_ERROR;
					break;
				case 'f':
				case 'F':
					error_state.breakpoint = ERROR_FATAL;
					break;
				case 'p':
				case 'P':
					error_state.breakpoint = ERROR_PANIC;
					break;
				default:
					error_state.breakpoint = strtoi(v, NULL, 0);
					break;
				}
			else
				error_state.breakpoint = 0;
			if (((Namval_t*)p)->value == OPT_CORE)
				error_info.core = error_state.breakpoint;
			break;
		case OPT_CATALOG:
			if (n)
				error_info.set |= ERROR_CATALOG;
			else
				error_info.clear |= ERROR_CATALOG;
			break;
		case OPT_COUNT:
			if (n)
				error_state.count = strtoul(v, NULL, 0);
			else
				error_state.count = 0;
			break;
		case OPT_FD:
			error_info.fd = n ? strtoi(v, NULL, 0) : -1;
			break;
		case OPT_LIBRARY:
			if (n)
				error_info.set |= ERROR_LIBRARY;
			else
				error_info.clear |= ERROR_LIBRARY;
			break;
		case OPT_MASK:
			if (n)
				error_info.mask = strtoi(v, NULL, 0);
			else
				error_info.mask = 0;
			break;
		case OPT_MATCH:
			if (error_state.match)
				regfree(error_state.match);
			if (n)
			{
				if ((error_state.match || (error_state.match = newof(0, regex_t, 1, 0))) && regcomp(error_state.match, v, REG_EXTENDED|REG_LENIENT))
				{
					free(error_state.match);
					error_state.match = 0;
				}
			}
			else if (error_state.match)
			{
				free(error_state.match);
				error_state.match = 0;
			}
			break;
		case OPT_PREFIX:
			if (n)
				error_state.prefix = strdup(v);
			else if (error_state.prefix)
			{
				free(error_state.prefix);
				error_state.prefix = 0;
			}
			break;
		case OPT_SYSTEM:
			if (n)
				error_info.set |= ERROR_SYSTEM;
			else
				error_info.clear |= ERROR_SYSTEM;
			break;
		case OPT_TIME:
			error_info.time = n ? 1 : 0;
			break;
		case OPT_TRACE:
			if (n)
				error_info.trace = -strtoi(v, NULL, 0);
			else
				error_info.trace = 0;
			break;
		}
	return 0;
}

/*
 * print a name with optional delimiter, converting unprintable chars
 */

static void
print(Sfio_t* sp, char* name, char* delim)
{
	if (mbwide())
		sfputr(sp, name, -1);
	else
	{
		/* the following code assumes ASCII */
		int		c;

		while (c = *name++)
		{
			if (c & 0200)
			{
				c &= 0177;
				sfputc(sp, '?');
			}
			if (c < ' ')
			{
				c += 'A' - 1;
				sfputc(sp, '^');
			}
			sfputc(sp, c);
		}
	}
	if (delim)
		sfputr(sp, delim, -1);
}

/*
 * print error context FIFO stack
 */

static void
context(Sfio_t* sp, Error_context_t* cp)
{
	if (cp->context)
		context(sp, cp->context);
	if (!(cp->flags & ERROR_SILENT))
	{
		if (cp->id)
			print(sp, cp->id, NULL);
		if (cp->line > ((cp->flags & ERROR_INTERACTIVE) != 0))
		{
			if (cp->file)
				sfprintf(sp, ": \"%s\", %s %d", cp->file, ERROR_translate(NULL, NULL, ast.id, "line"), cp->line);
			else
				sfprintf(sp, "[%d]", cp->line);
		}
		sfputr(sp, ": ", -1);
	}
}

/*
 * debugging breakpoint
 */

static void
error_break(void)
{
	char*	s;

	if (error_state.tty || (error_state.tty = sfopen(NULL, "/dev/tty", "r+")))
	{
		sfprintf(error_state.tty, "error breakpoint: ");
		if (s = sfgetr(error_state.tty, '\n', 1))
		{
			if (streq(s, "q") || streq(s, "quit"))
				exit(0);
			stropt(s, options, sizeof(*options), setopt, NULL);
		}
	}
}

COLD void
error(int level, ...)
{
	va_list	ap;

	va_start(ap, level);
	errorv(NULL, level, ap);
	va_end(ap);
}

COLD void
errorv(const char* id, int level, va_list ap)
{
	ptrdiff_t	n;
	int		fd;
	Sfio_t*		iop;
	int		flags;
	char*		s;
	char*		t;
	char*		format;
	char*		library;
	const char*	catalog;

	int		line;
	char*		file;

	unsigned long	d;
	struct tms	us;

	if (!error_info.init)
	{
		error_info.init = 1;
		stropt(getenv("ERROR_OPTIONS"), options, sizeof(*options), setopt, NULL);
	}
	if (level > 0)
	{
		flags = level & ~ERROR_LEVEL;
		level &= ERROR_LEVEL;
	}
	else
		flags = 0;
	if ((flags & (ERROR_USAGE|ERROR_NOID)) == ERROR_NOID)
	{
		format = (char*)id;
		id = 0;
	}
	else
		format = 0;
	if (id)
	{
		catalog = (char*)id;
		if (!*catalog || *catalog == ':')
		{
			catalog = 0;
			library = 0;
		}
		else if ((library = (char*)strchr(catalog, ':')) && !*++library)
			library = 0;
	}
	else
	{
		catalog = 0;
		library = 0;
	}
	if (catalog)
		id = 0;
	else
	{
		id = (const char*)error_info.id;
		catalog = error_info.catalog;
	}
	if (level < error_info.trace || (flags & ERROR_LIBRARY) && !(((error_info.set | error_info.flags) ^ error_info.clear) & ERROR_LIBRARY) || level < 0 && error_info.mask && !(error_info.mask & (1<<(-level - 1))))
	{
		if (level >= ERROR_FATAL)
			(*error_info.exit)(level - 1);
		return;
	}
	if (error_info.trace < 0)
		flags |= ERROR_LIBRARY|ERROR_SYSTEM;
	flags |= error_info.set | error_info.flags;
	flags &= ~error_info.clear;
	if (!library)
		flags &= ~ERROR_LIBRARY;
	iop = (flags & ERROR_SFIO_OUT) ? va_arg(ap, Sfio_t*) : NULL;
	fd = (flags & ERROR_OUTPUT) ? va_arg(ap, int) : error_info.fd;
	if (error_info.write)
	{
		ptrdiff_t	off;
		char*		bas;

		bas = stkptr(stkstd, 0);
		if (off = stktell(stkstd))
			stkfreeze(stkstd, 0);
		file = error_info.id;
		if (error_state.prefix)
			sfprintf(stkstd, "%s: ", error_state.prefix);
		if (flags & ERROR_USAGE)
		{
			if (flags & ERROR_NOID)
				sfprintf(stkstd, "       ");
			else
				sfprintf(stkstd, "%s: ", ERROR_translate(NULL, NULL, ast.id, "Usage"));
			if (file || opt_info.argv && (file = opt_info.argv[0]))
				print(stkstd, file, " ");
		}
		else
		{
			if (level && !(flags & ERROR_NOID))
			{
				if (error_info.context && level > 0)
					context(stkstd, error_info.context);
				if (file)
					print(stkstd, file, (flags & ERROR_LIBRARY) ? " " : ": ");
				if (flags & (ERROR_CATALOG|ERROR_LIBRARY))
				{
					sfprintf(stkstd, "[");
					if (flags & ERROR_CATALOG)
						sfprintf(stkstd, "%s %s%s",
							catalog ? catalog : ERROR_translate(NULL, NULL, ast.id, "DEFAULT"),
							ERROR_translate(NULL, NULL, ast.id, "catalog"),
							(flags & ERROR_LIBRARY) ? ", " : "");
					if (flags & ERROR_LIBRARY)
						sfprintf(stkstd, "%s %s",
							library,
							ERROR_translate(NULL, NULL, ast.id, "library"));
					sfprintf(stkstd, "]: ");
				}
			}
			if (level > 0 && error_info.line > ((flags & ERROR_INTERACTIVE) != 0))
			{
				if (error_info.file && *error_info.file)
					sfprintf(stkstd, "\"%s\", ", error_info.file);
				sfprintf(stkstd, "%s %d: ", ERROR_translate(NULL, NULL, ast.id, "line"), error_info.line);
			}
		}
		if (error_info.time)
		{
			if ((d = (unsigned long)times(&us)) < error_info.time || error_info.time == 1)
				error_info.time = d;
			sfprintf(stkstd, " %05lu.%05lu.%05lu ", d - error_info.time, (unsigned long)us.tms_utime, (unsigned long)us.tms_stime);
		}
		switch (level)
		{
		case 0:
			flags &= ~ERROR_SYSTEM;
			break;
		case ERROR_WARNING:
			sfprintf(stkstd, "%s: ", ERROR_translate(NULL, NULL, ast.id, "warning"));
			break;
		case ERROR_PANIC:
			sfprintf(stkstd, "%s: ", ERROR_translate(NULL, NULL, ast.id, "panic"));
			break;
		default:
			if (level < 0)
			{
				s = ERROR_translate(NULL, NULL, ast.id, "debug");
				if (error_info.trace < -1)
					sfprintf(stkstd, "%s%d:%s", s, level, level > -10 ? " " : "");
				else
					sfprintf(stkstd, "%s: ", s);
				for (n = 0; n < error_info.indent; n++)
				{
					sfputc(stkstd, ' ');
					sfputc(stkstd, ' ');
				}
			}
			break;
		}
		if (flags & ERROR_SOURCE)
		{
			/*
			 * source ([version], file, line) message
			 */

			file = va_arg(ap, char*);
			line = va_arg(ap, int);
			s = ERROR_translate(NULL, NULL, ast.id, "line");
			if (error_info.version)
				sfprintf(stkstd, "(%s: \"%s\", %s %d) ", error_info.version, file, s, line);
			else
				sfprintf(stkstd, "(\"%s\", %s %d) ", file, s, line);
		}
		if (format || (format = va_arg(ap, char*)))
		{
			if (!(flags & ERROR_USAGE))
				format = ERROR_translate(NULL, id, catalog, format);
			sfvprintf(stkstd, format, ap);
		}
		if (!(flags & ERROR_PROMPT))
		{
			/*
			 * level&ERROR_OUTPUT on return means message
			 * already output
			 */

			if ((flags & ERROR_SYSTEM) && errno && errno != error_info.last_errno)
			{
				sfprintf(stkstd, " [%s]", strerror(errno));
				if (error_info.set & ERROR_SYSTEM)
					errno = 0;
				error_info.last_errno = (level >= 0) ? 0 : errno;
			}
			if (error_info.auxiliary && level >= 0)
				level = (*error_info.auxiliary)(stkstd, level, flags);
			sfputc(stkstd, '\n');
		}
		if (level > 0)
		{
			if ((level & ~ERROR_OUTPUT) > 1)
				error_info.errors++;
			else
				error_info.warnings++;
		}
		if (level < 0 || !(level & ERROR_OUTPUT))
		{
			n = stktell(stkstd);
			s = stkptr(stkstd, 0);
			if (t = memchr(s, '\f', (size_t)n))
			{
				n -= ++t - s;
				s = t;
			}
#if HUH_19980401 /* nasty problems if sfgetr() is in effect! */
			sfsync(sfstdin);
#endif
			sfsync(sfstdout);
			sfsync(sfstderr);
			if (iop)
			{
				sfwrite(iop, s, (size_t)n);
				sfsync(iop);
			}
			else if (fd == sffileno(sfstderr) && error_info.write == write)
			{
				sfwrite(sfstderr, s, (size_t)n);
				sfsync(sfstderr);
			}
			else
				(*error_info.write)(fd, s, (size_t)n);
		}
		else
		{
			s = 0;
			level &= ERROR_LEVEL;
		}
		stkset(stkstd, bas, off);
	}
	else
		s = 0;
	if (level >= error_state.breakpoint && error_state.breakpoint && (!error_state.match || !regexec(error_state.match, s ? s : format, 0, NULL, 0)) && (!error_state.count || !--error_state.count))
	{
		if (error_info.core)
		{
#ifndef SIGABRT
#ifdef	SIGQUIT
#define SIGABRT	SIGQUIT
#else
#ifdef	SIGIOT
#define SIGABRT	SIGIOT
#endif
#endif
#endif
#ifdef	SIGABRT
			signal(SIGABRT, SIG_DFL);
			kill(getpid(), SIGABRT);
			pause();
#else
			abort();
#endif
		}
		else
			error_break();
	}
	if (level >= ERROR_FATAL)
		(*error_info.exit)(level - ERROR_FATAL + 1);
}
